torch/distributions/wishart.py

```diff
@@ -0,0 +1,262 @@
+        new._validate_args = self._validate_args
+        return new
+
+    @lazy_property
+    def scale_tril(self):
+        return self._unbroadcasted_scale_tril.expand(
+            self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def covariance_matrix(self):
+        return (
+            self._unbroadcasted_scale_tril @ self._unbroadcasted_scale_tril.transpose(-2, -1)
+        ).expand(self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def precision_matrix(self):
+        identity = torch.eye(
+            self._event_shape[-1],
+            device=self._unbroadcasted_scale_tril.device,
+            dtype=self._unbroadcasted_scale_tril.dtype,
+        )
+        return torch.cholesky_solve(
+            identity, self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @property
+    def mean(self):
+        return self.df.view(self._batch_shape + (1, 1,)) * self.covariance_matrix
+
+    @property
+    def variance(self):
+        V = self.covariance_matrix  # has shape (batch_shape x event_shape)
+        diag_V = V.diagonal(dim1=-2, dim2=-1)
+        return self.df.view(self._batch_shape + (1, 1,)) * (V.pow(2) + torch.einsum("...i,...j->...ij", diag_V, diag_V))
+
+    def _bartlett_sampling(self, sample_shape=torch.Size()):
+        p = self._event_shape[-1]  # has singleton shape
+
+        # Implemented Sampling using Bartlett decomposition
+        noise = self._dist_chi2.rsample(sample_shape).sqrt().diag_embed(dim1=-2, dim2=-1)
+        i, j = torch.tril_indices(p, p, offset=-1)
+        noise[..., i, j] = torch.randn(
+            torch.Size(sample_shape) + self._batch_shape + (p * (p - 1) // 2,),
+            dtype=noise.dtype,
+            device=noise.device,
+        )
+        chol = self._unbroadcasted_scale_tril @ noise
+        return chol @ chol.transpose(-2, -1)
+
+    def rsample(self, sample_shape=torch.Size(), max_try=10):
+        sample_shape = torch.Size(sample_shape)
+        sample = self._bartlett_sampling(sample_shape)
+
+        # Below part is to improve numerical stability temporally and should be removed in the future
+        is_singular = self.support.check(sample).logical_not()
+        if self._batch_shape:
+            is_singular = is_singular.amax(self._reduced_batch_dims)
+
+        if is_singular.any():
+            warnings.warn("Singular sample detected.")
+
+            for _ in range(max_try):
+                if sample_shape or not is_singular.all():
+                    new_sample = self._bartlett_sampling(is_singular[is_singular].shape)
+                    sample = sample[is_singular.logical_not()]
+
+                    is_singular = self.support.check(new_sample).logical_not()
+                    if self._batch_shape:
+                        is_singular = is_singular.amax(self._reduced_batch_dims)
+
+                    sample = torch.cat(
```

Comment:
Won't this destroy the indexing in `is_singular`. e.g. index 3 is singular but we concatenate a valid sample here at index 3 which will cause issues in the next iteration. I just ran this for the following case, and it gives some errors:

```python
w = Wishart(df=torch.tensor(1.1), scale_tril=torch.randn(3, 2, 2).pow(2.).tril())
w.rsample((5,))
```

It may be worth keeping the shape of `sample` and `is_singular` the same throughout. How about something like the following? (not tested)

```
n_try = 0
while n_try < max_try:
    new_sample = self._bartlett_sampling(is_singular[is_singular].shape)
    sample[is_singular] = new_sample
    is_singular = ~self.support.check(sample)
    if self._batch_shape:
        is_singular = is_singular.amax(self._reduced_batch_dims)

    if not is_singular.any():
        break
    n_try += 1
```

Reply:
I will test your code. Thanks for your effort.

Reply:
`~self.support.check(sample)` seems redundant, but your code is much simpler. I will go on with your implementation.

Reply:
I agree, that's a good point! What do you think about doing something like the following to avoid redundant checks:

```python
  is_singular_new = ~self.support.check(new_sample)
  if self._batch_shape:
      is_singular_new = is_singular_new.amax(self._reduced_batch_dims)
  is_singular[is_singular] = is_singular_new
```

Reply:
That would be great. I will apply your suggestions.